fv3core/pace/fv3core/initialization/init_utils.py

```diff
@@ -0,0 +1,414 @@
+        lon[1:, 0:-1], lon[1:, 1:], lat[1:, 0:-1], lat[1:, 1:], np
+    )
+    _, lat_avg_x_north = lon_lat_midpoint(
+        lon[0:-1, 1:], lon[1:, 1:], lat[0:-1, 1:], lat[1:, 1:], np
+    )
+    _, lat_avg_y_west = lon_lat_midpoint(
+        lon[:, 0:-1], lon[:, 1:], lat[:, 0:-1], lat[:, 1:], np
+    )
+    return lat_avg_x_south, lat_avg_y_east, lat_avg_x_north, lat_avg_y_west
+
+
+def compute_slices(nx, ny):
+    islice = slice(NHALO, NHALO + nx)
+    jslice = slice(NHALO, NHALO + ny)
+    slice_3d = (islice, jslice, slice(None))
+    slice_2d = (islice, jslice)
+    return islice, jslice, slice_3d, slice_2d
+
+
+def empty_numpy_dycore_state(shape):
+    numpy_dict = {}
+    for _field in fields(DycoreState):
+        if "dims" in _field.metadata.keys():
+            numpy_dict[_field.name] = np.zeros(
+                shape[: len(_field.metadata["dims"])],
+                dtype=Float,
+            )
+    numpy_state = SimpleNamespace(**numpy_dict)
+    return numpy_state
+
+
+def _find_midpoint_unit_vectors(p1, p2):
+
+    midpoint = np.array(
+        lon_lat_midpoint(p1[:, :, 0], p2[:, :, 0], p1[:, :, 1], p2[:, :, 1], np)
+    ).transpose([1, 2, 0])
+    unit_dir = get_unit_vector_direction(p1, p2, np)
+    exv, eyv = get_lonlat_vect(midpoint, np)
+
+    muv = {"midpoint": midpoint, "unit_dir": unit_dir, "exv": exv, "eyv": eyv}
+
+    return muv
+
+
+def fix_top_log_edge_pressure(peln, ptop):
+    if ptop < PTOP_MIN:
+        ak1 = (constants.KAPPA + 1.0) / constants.KAPPA
+        peln[:, :, 0] = peln[:, :, 1] - ak1
+    else:
+        peln[:, :, 0] = np.log(ptop)
+
+
+def geopotential_perturbation(lat, eta_value):
+    """
+    Equation (7) JRMS2006, just the perturbation component
+    """
+    u_comp = U0 * (np.cos(eta_value) ** (3.0 / 2.0))
+    return u_comp * (
+        (-2.0 * (np.sin(lat) ** 6.0) * (np.cos(lat) ** 2.0 + 1.0 / 3.0) + 10.0 / 63.0)
+        * u_comp
+        + (
+            (8.0 / 5.0) * (np.cos(lat) ** 3.0) * (np.sin(lat) ** 2.0 + 2.0 / 3.0)
+            - math.pi / 4.0
+        )
+        * constants.RADIUS
+        * constants.OMEGA
+    )
+
+
+def horizontally_averaged_temperature(eta):
+    """
+    Equations (4) and (5) JRMS2006 for characteristic temperature profile
+    """
+    # for troposphere:
+    t_mean = T_0 * eta[:] ** (constants.RDGAS * LAPSE_RATE / constants.GRAV)
+    # above troposphere
+    t_mean[ETA_TROPOPAUSE > eta] = (
+        t_mean[ETA_TROPOPAUSE > eta]
+        + DELTA_T * (ETA_TROPOPAUSE - eta[ETA_TROPOPAUSE > eta]) ** 5.0
+    )
+    return t_mean
+
+
+def _initialize_delp(ak, bk, ps, shape):
+    delp = np.zeros(shape)
+    delp[:, :, :-1] = (
+        ak[None, None, 1:]
+        - ak[None, None, :-1]
+        + ps[:, :, None] * (bk[None, None, 1:] - bk[None, None, :-1])
+    )
+
+    return delp
+
+
+def initialize_delp(ps, ak, bk):
+    return (
+        ak[None, None, 1:]
+        - ak[None, None, :-1]
+        + ps[:, :, None] * (bk[None, None, 1:] - bk[None, None, :-1])
+    )
```

Comment:
This duplication is silly, we should consolidate at some point

Reply:
I will add this into the work I will do after this PR